### DISCUSSION TITLE
How to declare the types definitions `.types.d.ts` in VSCode/Sublime Text

### DISCUSSION THREAD

**Discussion by darkred** [tophf/mpiv#50] [MPIV general]
Greetings @tophf 

I'm trying to declare the types definitions in https://github.com/tophf/mpiv/blob/master/.types.d.ts .
The editors I use are latest VSCode and Sublime Text 3
*(the former has TypeScript language support but does not include the TypeScript compiler, `tsc` --- the latter needs a package to offer support)* .
(Unfortunately I couldn't find any info in https://definitelytyped.org/directory/tools.html, nor anything related to "userscripts and .types.d.ts" in general).

As I see in this [SO answer](https://stackoverflow.com/a/61664283) I need a `tsconfig.json` .
Could you please  share your `tsconfig.json` ? (and maybe upload it in the repo?) And, should I do anything else in order for it to work at least in VSCode?
Or, if you don't use these editors, could you offer some info on the editor you use?

Thank you

**Comment by tophf** [comment]
I don't have a tsconfig though, that file is just automatically recognized as a part of the project so I have no idea how to replicate it, I don't even use TypeScript, it was just for fun. If you find the proper method, just open a PR.

**Reply by darkred** [reply]
Thank you for the quick answer. If I find a way I'll surely open a PR.

>  that file is just automatically recognized as a part of the project 

So, for you it's working in your editor/IDE, as it is, i.e. without any configuration changes. 
What editor/IDE yo you use? VSCode? Visual Studio?

**Reply by tophf** [reply]
Mine is a commercial one so it's irrelevant. Try pmneo.tsimporter extension in VSCode or maybe there's another extension.

**Reply by darkred** [reply]
According to https://stackoverflow.com/a/53015835/3231411 :

> 2018 now.  You don't need any extensions for auto-imports in Javascript (as long as you have `checkjs: true` in your `jsconfig.json` file) and TypeScript.

therefore adding a `jsconfig.json` like the following in the same folder with the script, should work ok in VSCode:

```json
{
	"compilerOptions": {
		"target": "es2018",
		"checkJs": true,
	},
	"include": ["*"]
}
```

But, the thing is that, almost all types declared in `.types.d.ts` are not present in the script, apart only from `Config` 
*(plus `HostRule`, `RuleMatchInfo`, and `AppInfo` which are only present in JSDoc comments `/** @returns @type */`)* .

As you said, you had added it just for fun.
So, maybe it's more preferable to just remove the `.types.d.ts` file from the repo, to avoid confusion? 
Or maybe update it, removing the non existing types, and adding the current ones?
And, should I make a PR with the above info I found?

**Reply by tophf** [reply]
Why remove it? It's still fun using it as it works for me :-)

**Reply by tophf** [reply]
BTW do you actually know TS? All types are used by each other there, at least this is how I intended it.

**Reply by darkred** [reply]
> Why remove it?

> All types are used by each other there, at least this is how I intended it.

Of all the 16 types declared in there: 
- only 1 type is actually used in the script, `Config`.
- 3 are found only in JSDoc comments,
- for the rest 12 I did a search e.g. for `PopupInfo` or `GalleryItem` and there were no matches.

Therefore all but 1 type should all be removed, if I'm not mistaken.

---

> BTW do you actually know TS? 

I know a few things about TS. I wanted to make use of that TS types file, because you use it: you are much more experienced than me, and  I'd to learn and follow your way/tools of programming.

**Reply by tophf** [reply]
JSDoc comments are used by IDE so all the types are used. Note that many of them are supplementary types used by other types.

**Reply by darkred** [reply]
Ok. Thanks for your time and info.

**Reply by darkred** [reply]
Hi, @tophf , I opened this PR #53
